flask_test_app/app.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Flask application made for integration tests."""
+
+import time
+from urllib.parse import urlparse
+
+import psycopg
+import pymysql
+import pymysql.cursors
+from flask import Flask, g, jsonify, request
+
+app = Flask(__name__)
+app.config.from_prefixed_env()
+
+
+def get_mysql_database():
+    """Get the mysql db connection."""
+    if "mysql_db" not in g:
+        if app.config.get("MYSQL_DB_CONNECT_STRING"):
```

Comment:
Shouldn't this be an environment variable?

Reply:
I don't understand: this is an environment variable processed by flask ?

Reply:
I suppose the question is whether this should be processed by flask or if it should be an environment variable that doers not get picked up by flask configuration. The reason is that the database connection string contains sensitive information and hence generally shouldn't be in configuration and instead be passed using just environment variables

Reply:
Similar to that sensitive values in a terraform file are passed using a vault rather than written into the terraform file

Reply:
For now, I'm strictly complying to the spec. This should be decided at the spec level.

Reply:
From my interpretation that line means that it shouldn't end up in the flask configuration, it should just be an environment variable in the container and that was my intent when writing the spec

Reply:
I'm not sure to understand how the flask application can configure its connection to a db without access to the db_uri. And the availability to flask of a env variable does not differ if it's processed by flask's config mechanism (`os.environ` is always available). What is the threat model we're are addressing by making it not available through flask's config ?

Reply:
Alright, good points, just trying to figure out whether I understand things correctly:

* looking at the code, my understanding is that flask is constructing a config from environment variables
* that config is not persisted

Now if these two are valid, the proposal is to not have flask know where to connect to via direct access to an environment variable, rather a middleware would get that info (from an environment variable) and would present it to flask, correct? I'm wondering what is the benefit here.
If there are security improvements, they should be noted - to my understanding you will always end up with accessing env variables and keeping them in volatile memory (again, correct me if I'm wrong here). If there are no security improvements, following flask spec might be a good thing - even if we give a PaaS solution, twisting the design of a Flask app from something expected (and documented) to something else should be well reasoned about and well documented.

Reply:
After discussion with David, it was decided that I should go the `os.environ` route for now and that more discussions about it could arise.

Reply:
I am a little bit confused by the risk introduced by putting the database URI string into the flask configuration. I think if an attacker can access the flask configuration, the attacker should have access to the environment variable too since they are all unprotected in the memory.

Reply:
To add to the discussion: even in flask documentation, there are examples of using `app.config` to store secrets: https://flask.palletsprojects.com/en/2.3.x/config/#configuring-from-environment-variables